helix-lsp/src/lib.rs

```diff
@@ -263,83 +307,140 @@ pub mod util {
     pub fn generate_transaction_from_snippet(
         doc: &Rope,
         selection: &Selection,
-        start_offset: i128,
-        end_offset: i128,
+        start_offset: Option<i128>,
+        end_offset: Option<i128>,
         snippet: snippet::Snippet,
         line_ending: &str,
         include_placeholder: bool,
+        tab_width: usize,
     ) -> Transaction {
         let text = doc.slice(..);
+        let mut new_selection: SmallVec<[_; 1]> = SmallVec::new();
 
-        // For each cursor store offsets for the first tabstop
-        let mut cursor_tabstop_offsets = Vec::<SmallVec<[(i128, i128); 1]>>::new();
-        let transaction = Transaction::change_by_selection(doc, selection, |range| {
-            let cursor = range.cursor(text);
-            let replacement_start = (cursor as i128 + start_offset) as usize;
-            let replacement_end = (cursor as i128 + end_offset) as usize;
-            let newline_with_offset = format!(
-                "{line_ending}{blank:width$}",
-                line_ending = line_ending,
-                width = replacement_start - doc.line_to_char(doc.char_to_line(replacement_start)),
-                blank = ""
-            );
-
-            let (replacement, tabstops) =
-                snippet::render(&snippet, newline_with_offset, include_placeholder);
-
-            let replacement_len = replacement.chars().count();
-            cursor_tabstop_offsets.push(
-                tabstops
-                    .first()
-                    .unwrap_or(&smallvec![(replacement_len, replacement_len)])
-                    .iter()
-                    .map(|(from, to)| -> (i128, i128) {
-                        (
-                            *from as i128 - replacement_len as i128,
-                            *to as i128 - replacement_len as i128,
-                        )
-                    })
-                    .collect(),
-            );
-
-            (replacement_start, replacement_end, Some(replacement.into()))
-        });
+        let (transaction, primary_idx) = Transaction::change_by_selection_ignore_overlapping(
+            doc,
+            selection,
+            |range, changeset| {
+                let cursor = range.cursor(text);
+                let (replacement_start, replacement_end) =
+                    completion_pos(text, start_offset, end_offset, cursor)?;
+                // Asssoc::After is used here because there should be no edits after
+                // `replacement_start` (those haven't been generated yet)
+                let mapped_replacement_start = changeset.map_pos(replacement_start, Assoc::After);
+                // if a snippet needs access to it's position (line number or offset) we must apply
+                // previous edits to the rope, as this has some ovherhead we do it lazily
+                // TODO: can we avoid apply the previous edits so this becomes O(N) instead of
+                // O(NlogN) with the number of selection ranges?
```

Comment:
Pretty sure it doesn't, because applying changeet is O(N) but the cahgneset only has the length of the already processed changes so during the iteration you are applying an O(N) N=1, O(N), N=2,,, O(N) N=M transaction. That's a classical example of an O(MlogM) algorithm. Still not great and just head an idea to optimize this further so I might be able to make it O(N)

Reply:
1 + 2 + ... + M = M^2 / 2 + M / 2 :)

Reply:
the new rope is generated right inside the snippets now eagerly the overhead should be quite small now. Also made the code simpler.